confidant/routes/v1.py

```diff
@@ -324,18 +366,117 @@ def map_service_credentials(id):
-    added = list(set(service.credentials) - set(_service_credential_ids))
-    removed = list(set(_service_credential_ids) - set(service.credentials))
-    msg = 'Added credentials: {0}; Removed credentials {1}; Revision {2}'
-    msg = msg.format(added, removed, service.revision)
-    graphite.send_event([id], msg)
+    servicemanager.send_service_mapping_graphite_event(service, _service)
     webhook.send_event('service_update', [service.id], service.credentials)
+    credentials = credentialmanager.get_credentials(
+        service.credentials,
+        metadata_only=True,
+    )
+    blind_credentials = credentialmanager.get_blind_credentials(
+        service.blind_credentials,
+        metadata_only=True,
+    )
+    return jsonify({
+        'id': service.id,
+        'account': service.account,
+        'credentials': credentials,
+        'blind_credentials': blind_credentials,
+        'revision': service.revision,
+        'enabled': service.enabled,
+        'modified_date': service.modified_date,
+        'modified_by': service.modified_by
+    })
+
+
+@app.route('/v1/services/<id>/<revision>', methods=['PUT'])
+@authnz.require_auth
+@authnz.require_csrf_token
+@maintenance.check_maintenance_mode
+def revert_service_to_revision(id, revision):
     try:
-        credentials = credentialmanager.get_credentials(service.credentials)
-    except KeyError:
-        return jsonify({'error': 'Decryption error.'}), 500
+        current_service = Service.get(id)
+        if current_service.data_type != 'service':
+            msg = 'id provided is not a service.'
+            return jsonify({'error': msg}), 400
+        new_revision = servicemanager.get_latest_service_revision(
+            id,
+            current_service.revision
+        )
+    except DoesNotExist:
+        logging.warning(
+            'Item with id {0} does not exist.'.format(id)
+        )
+        return jsonify({}), 404
+    try:
+        revert_service = Service.get('{}-{}'.format(id, revision))
+        if revert_service.data_type != 'archive-service':
+            msg = 'id provided is not a service.'
+            return jsonify({'error': msg}), 400
+    except DoesNotExist:
+        logging.warning(
+            'Item with id {0} does not exist.'.format(id)
+        )
+        return jsonify({}), 404
+    if revert_service.equals(current_service):
+        ret = {
+            'error': 'No difference between old and new service.'
+        }
+        return jsonify(ret), 400
+    if revert_service.credentials or revert_service.blind_credentials:
+        conflicts = credentialmanager.pair_key_conflicts_for_credentials(
+            revert_service.credentials,
+            revert_service.blind_credentials,
+        )
+        if conflicts:
+            ret = {
+                'error': 'Conflicting key pairs in mapped service.',
+                'conflicts': conflicts
+            }
+            return jsonify(ret), 400
+    # Try to save to the archive
```

Comment:
Curious -- how come we save to archive here?   Is it a sort of backup?

Reply:
Confidant's dynamo usage is a bit weird. We use a single table for all record types, and we store history and current records in the same table. For performance, we have two types of record per model. So here, we have `service` and `service-archive` type records. The `service` record is the one that's queried, when `get_service` is called. When history is looked up, it's the archive records.

If we were to split the records into separate tables, we would use hash+key, which would let us drop the double-save. It's something I've considered, if we were to get rid of blind credentials (which we can do with proper RBAC).